src/libsyntax/parse/parser.rs

```diff
@@ -572,9 +572,13 @@ impl<'a> Parser<'a> {
                 self.bug("ident interpolation not converted to real token");
             }
             _ => {
-                let token_str = self.this_token_to_string();
-                Err(self.fatal(&format!("expected ident, found `{}`",
-                                    token_str)))
+                let mut err = self.fatal(&format!("expected identifier, found `{}`",
+                                                  self.this_token_to_string()));
+                if self.token == token::Underscore {
+                    err.fileline_note(self.span, "`_` is not an identifier but a wildcard pattern \
+                                                  and patterns are not expected in this position");
```

Comment:
I think this second half of the message could be omitted. It's a bit awkward to repeat what's expected here. The important piece of information that many users are missing is that `_` is not an identifier. "expected identifier" covers that it's not legal in this position.